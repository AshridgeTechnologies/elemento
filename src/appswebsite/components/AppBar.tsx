import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}