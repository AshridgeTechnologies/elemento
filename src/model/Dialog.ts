import Element from './Element'
import BaseElement, {propDef} from './BaseElement'
import {ComponentType, ElementType, ParentType, PropertyDef, PropertyValueType, Styling} from './Types'
import {elementHasParentTypeOf} from './createElement'
import {BlockLayout, blockLayoutChoices} from './Block'

type Properties = Partial<Readonly<{layout: BlockLayout, initiallyOpen: PropertyValueType<boolean>, showCloseButton: PropertyValueType<boolean>}>> & Styling

export default class Block extends BaseElement<Properties> implements Element {

    readonly kind = 'Dialog'
    get iconClass() { return 'table_view' }
    type(): ComponentType { return 'statefulUI' }

    get layout() { return this.properties.layout }
    get initiallyOpen() { return this.properties.initiallyOpen }
    get showCloseButton() { return this.properties.showCloseButton }
    get styles() { return this.properties.styles }

    get propertyDefs(): PropertyDef[] {
        return [
            propDef('layout', blockLayoutChoices),
            propDef('initiallyOpen', 'boolean', {state: true}),
            propDef('showCloseButton', 'boolean'),
            propDef('styles', 'styles'),
        ]
    }

    canContain(elementType: ElementType) {
        return elementHasParentTypeOf(elementType, this)
    }

    static get parentType(): ParentType { return 'Page' }
}
