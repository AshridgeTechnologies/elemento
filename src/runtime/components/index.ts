export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
