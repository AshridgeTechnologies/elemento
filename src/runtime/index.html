<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>