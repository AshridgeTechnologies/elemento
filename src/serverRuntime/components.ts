export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts