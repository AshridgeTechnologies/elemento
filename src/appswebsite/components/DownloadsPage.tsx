import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
