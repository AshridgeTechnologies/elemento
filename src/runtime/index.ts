import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
