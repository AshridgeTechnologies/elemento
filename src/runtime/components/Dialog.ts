import React from 'react'
import {StylesPropVals, valueOfProps} from '../runtimeFunctions'
import {Dialog as MuiDialog, IconButton} from '@mui/material'
import {BaseComponentState, ComponentState} from './ComponentState'
import {BlockContent, BlockLayout} from './Block'
import Close from '@mui/icons-material/Close'
import {use$state} from '../state/appStateHooks'
import {ElementMetadata, ElementSchema} from '../../model/ModelElement'
import {Definitions} from '../../model/schema'

type Properties = { path: string, layout: BlockLayout, initiallyOpen?: boolean, showCloseButton?: boolean, styles?: StylesPropVals, children?: React.ReactElement[] }
type StateProperties = Partial<Readonly<{initiallyOpen: boolean}>>

type StateUpdatableProperties = Partial<Readonly<{ isOpen: boolean }>>

const DialogSchema: ElementSchema = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Dialog",
    "description": "A dialog box that can be opened and closed.  It can contain any elements, and can be given a title. ",
    "type": "object",
    "$ref": "#/definitions/BaseElement",
    "kind": "Dialog",
    "icon": "table_view",
    "elementType": "statefulUI",
    "canContain": "elementsWithThisParentType",
    "parentType": "Page",
    "properties": {
        "properties": {
            "type": "object",
            "unevaluatedProperties": false,
            "properties": {
                "layout": {
                    "description": "The ",
                    "enum": ["vertical", "horizontal", "horizontal wrapped", "positioned", "none"]
                },
                "initiallyOpen": {
                    "description": "The ",
                    "$ref": "#/definitions/BooleanOrExpression"
                },
                "showCloseButton": {
                    "description": "The ",
                    "$ref": "#/definitions/BooleanOrExpression"
                },
                "styles": {
                    "description": "The ",
                    "$ref": "#/definitions/Styles"
                }
            }
        },
        "elements": {
            "type": "array",
            "items": {
                "$ref": "#/definitions/BaseElement"
            }
        }
    },
    "required": ["kind", "properties"],
    "unevaluatedProperties": false,
    "definitions": Definitions
}

export const DialogMetadata: ElementMetadata = {
    stateProps: ['initiallyOpen']
}

const dialogSlotProps = {
    root: {
        sx: {position: 'absolute'},
    },
    backdrop: {
        sx: {
            position: 'absolute',
            backgroundColor: 'rgba(0, 0, 0, 0.25)'
        }}

}

function CloseButton(props: { onClose: () => void }) {
    const iconButtonProps = {edge: 'start', color: 'inherit', onClick: props.onClose, 'aria-label': 'close', 'data-testid': 'dialog_close_button',
        sx:{position: 'absolute', right: 0, top: 0}}
    // @ts-ignore
    return React.createElement(IconButton, iconButtonProps,
        React.createElement(Close)
    )
}

export default function Dialog({children = [], path,  layout, initiallyOpen, showCloseButton, styles: styleProps = {}}: Properties) {
    const styles = valueOfProps(styleProps)
    const state = use$state(path, DialogState, {initiallyOpen})
    const handleClose = ()=> { state.Close() }
    return React.createElement(MuiDialog, {open: state.isOpen, onClose: handleClose, slotProps: dialogSlotProps},
        (showCloseButton ? React.createElement(CloseButton, {onClose: handleClose}) : null),
        React.createElement(BlockContent, {path, layout, styles}, children)
    )
}
export class DialogState extends BaseComponentState<StateProperties, StateUpdatableProperties>
    implements ComponentState<DialogState> {

    get isOpen() {
        return this.state.isOpen ?? this.props.initiallyOpen ?? false
    }

    Show() {
        if (!this.latest().isOpen) {
            this.updateState({isOpen: true})
        }
    }

    Close() {
        if (this.latest().isOpen) {
            this.updateState({isOpen: false})
        }
    }
}

Dialog.State = DialogState
Dialog.Schema = DialogSchema
Dialog.Metadata = DialogMetadata
