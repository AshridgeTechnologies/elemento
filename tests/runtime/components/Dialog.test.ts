/**
 * @jest-environment jsdom
 */

import {createElement} from 'react'
import {Dialog, TextElement} from '../../../src/runtime/components'
import {wrappedTestElement} from '../../testutil/testHelpers'
import {DialogState} from '../../../src/runtime/components/Dialog'
import {actWait} from '../../testutil/rtlHelpers'
import {fireEvent, render, within} from '@testing-library/react'
import '@testing-library/jest-dom'

const [dialog, appStoreHook] = wrappedTestElement(Dialog, DialogState)

const stateAt = (path: string) => appStoreHook.stateAt(path)

const text1 = createElement(TextElement, {path: 'app.page1.things.text1', styles: {width: 200, top: 20, left: 30}, content: 'First text'} )
const text2 = createElement(TextElement, {path: 'app.page1.things.text2', styles: {width: 300, bottom: 10, right: 34}, content: 'Second text'} )

test('Dialog element shows dialog when initially closed', () => {
    const element = dialog('app.page1.confirmIt', {}, {styles: {color: 'red'}}, text1, text2)
    const { baseElement } = render(element)
    expect(baseElement).toMatchSnapshot();
})

test('Dialog element shows dialog when initially open',() => {
    const element = dialog('app.page1.confirmIt', {initiallyOpen: true}, {layout: 'vertical', styles: {color: 'red'}}, text1)
    const { baseElement } = render(element)
    expect(baseElement).toMatchSnapshot()
})

test('Dialog can be closed and opened via state object', async () => {
    let queryByText: any
    const element = dialog('app.page1.confirmIt', {}, {}, text1)
    await actWait( () => ({queryByText} = render(element)))
    await actWait( () => appStoreHook.setStateAt('app.page1.confirmIt', new DialogState({})))

    expect(queryByText('First text')).toBe(null)
    expect(stateAt('app.page1.confirmIt').isOpen).toBe(false)

    await actWait( () => stateAt('app.page1.confirmIt').Show() )
    expect(queryByText('First text')).not.toBe(null)
    expect(stateAt('app.page1.confirmIt').isOpen).toBe(true)

    await actWait( () => stateAt('app.page1.confirmIt').Close() )
    expect(queryByText('First text')).not.toBeVisible()
    expect(stateAt('app.page1.confirmIt').isOpen).toBe(false)
})

test('Dialog can be closed by close button', async () => {
    let queryByText: any, getByTestId: any
    const element = dialog('app.page1.confirmIt', {initiallyOpen: true}, {showCloseButton: true}, text1)
    await actWait( () => ({queryByText, getByTestId} = render(element)))

    expect(queryByText('First text')).not.toBe(null)

    await actWait( () => fireEvent.click(getByTestId('dialog_close_button')))
    expect(queryByText('First text')).not.toBeVisible()
    expect(stateAt('app.page1.confirmIt').isOpen).toBe(false)
})
