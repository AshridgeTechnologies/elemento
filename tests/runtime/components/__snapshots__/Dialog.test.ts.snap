// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog element shows dialog when initially closed 1`] = `
<body>
  <div />
</body>
`;

exports[`Dialog element shows dialog when initially open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-1u7a867-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-9m8e46-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-19do60a-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r1:"
        aria-modal="true"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-10d30g3-MuiPaper-root-MuiDialog-paper"
        role="dialog"
        style="--Paper-shadow: 0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiStack-root css-xmd97m-MuiStack-root"
          id="app.page1.confirmIt"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 css-1ewvver-MuiTypography-root"
            id="app.page1.things.text1"
          >
            First text
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
